test/validation/api/scheduler/scheduler.c

```diff
@@ -2348,6 +2347,130 @@ static void scheduler_test_order_wait_1_thread(void)
+	while ((ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT))
+	       != ODP_EVENT_INVALID) {
+		/* We got an event that was meant for the main thread */
+		odp_event_free(ev);
+	}
+
+	return 0;
+}
+
+static void scheduler_test_order_wait(void)
+{
+	odp_schedule_capability_t sched_capa;
+	odp_queue_param_t queue_param;
+	odp_queue_t queue;
+	pthrd_arg thr_arg = {.numthrds = 1};
+	int ret;
+	odp_event_t ev;
+
+	CU_ASSERT(!odp_schedule_capability(&sched_capa));
+
+	sched_queue_param_init(&queue_param);
+	queue_param.sched.sync = ODP_SCHED_SYNC_ORDERED;
+	queue = odp_queue_create("ordered queue", &queue_param);
+	CU_ASSERT_FATAL(queue != ODP_QUEUE_INVALID);
+	CU_ASSERT_FATAL(odp_queue_type(queue) == ODP_QUEUE_TYPE_SCHED);
+	CU_ASSERT_FATAL(odp_queue_sched_type(queue) == ODP_SCHED_SYNC_ORDERED);
+
+	odp_atomic_init_u32(&globals->order_wait.helper_ready, 0);
+	odp_atomic_init_u32(&globals->order_wait.helper_active, 0);
+
+	ret = odp_cunit_thread_create(order_wait_helper, &thr_arg);
+	CU_ASSERT_FATAL(ret == thr_arg.numthrds);
+
+	/* Send an event to the helper thread */
+	enqueue_event(queue);
+
+	/* Wait that the helper thread gets the event */
+	while (!odp_atomic_load_acq_u32(&globals->order_wait.helper_ready))
+		;
+
+	/* Try to send an event to ourselves */
+	enqueue_event(queue);
+	/*
+	 * If ordered queues are implemented as atomic queues, the schedule
+	 * call here will not return anything until the helper thread has
+	 * released the scheduling context of the first event. So we have
+	 * to wait long enough before giving up.
+	 */
+	ev = odp_schedule(NULL, odp_schedule_wait_time(2 * ODP_TIME_SEC_IN_NS));
+	if (ev == ODP_EVENT_INVALID) {
+		/* Helper thread got the event. Give up. */
+		printf("SKIPPED...");
+		goto out;
+	}
+	odp_event_free(ev);
+
+	/*
+	 * We are now in an ordered scheduling context and behind the helper
+	 * thread in source queue order if the helper thread has not released
+	 * the scheuduling context.
+	 */
+
+	if (!odp_atomic_load_acq_u32(&globals->order_wait.helper_active)) {
+		/*
+		 * Helper thread has released the context already.
+		 * We cannot test order wait fully.
+		 */
+		printf("reduced test...");
+	}
+
+	/*
+	 * The function we are testing: Wait until there are no scheduling
+	 * contexts that precede ours.
+	 */
+	odp_schedule_order_wait();
+
+	/*
+	 * If order wait is supported, we are now first in the source queue
+	 * order, so the helper thread must have released its context.
+	 */
+	if (sched_capa.order_wait)
+		CU_ASSERT(!odp_atomic_load_acq_u32(&globals->order_wait.helper_active));
+
+	/* Release the context */
+	ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+	CU_ASSERT(ev == ODP_EVENT_INVALID);
+
+out:
+	CU_ASSERT(odp_cunit_thread_exit(&thr_arg) == 0);
+	CU_ASSERT(odp_queue_destroy(queue) == 0);
+}
+
+static int check_2_workers(void)
+{
+	if (globals->num_workers < 2) {
+		printf("\nTest: scheduler_test_order_wait: SKIPPED\n");
+		return ODP_TEST_INACTIVE;
+	}
+
+	return ODP_TEST_ACTIVE;
+}
```

Comment:
Is there a need to run `scheduler_test_order_wait()` if `sched_capa.order_wait` is not set?

Reply:
Perhaps not strictly necessary, but the API says that the function does nothing in that case (i.e. it can still be called even though it is not very useful), so the test calls it even if the capa is not set.

Reply:
OK. Angle was that as it's already called in `scheduler_test_order_wait_1_thread()` unconditionally, the new test could be just skipped if there's no capa.

Reply:
I think doing the call in the new test too even when the capa is not there does not harm and might even expose a problem since the new test has two threads running. Plus then the capability function gets called too. So I think I will keep it.